Sources/PackageSyntax/ManifestRewriter.swift

```diff
@@ -0,0 +1,822 @@
+        let newManifest = targetDependencies.withAdditionalElementExpr(ExprSyntax(
+            StringLiteralExprSyntax(dependency)
+        )).root
+
+        self.editedSource = newManifest.as(SourceFileSyntax.self)!
+    }
+
+    public func addProductTargetDependency(
+        target: String,
+        product: String,
+        package: TextualPackageReference
+    ) throws {
+        let targetDependencies = try findTargetDependenciesArrayExpr(target: target)
+
+        let dotProductExpr = MemberAccessExprSyntax(base: nil,
+                                                                dot: TokenSyntax.periodToken(),
+                                                                name: TokenSyntax.identifier("product"),
+                                                                declNameArguments: nil)
+
+        let argumentList = TupleExprElementListSyntax([
+            TupleExprElementSyntax(label: TokenSyntax.identifier("name"),
+                                               colon: TokenSyntax.colonToken(trailingTrivia: .spaces(1)),
+                                               expression: ExprSyntax(StringLiteralExprSyntax(product)),
+                                               trailingComma: TokenSyntax.commaToken(trailingTrivia: .spaces(1))),
+            TupleExprElementSyntax(label: TokenSyntax.identifier("package"),
+                                               colon: TokenSyntax.colonToken(trailingTrivia: .spaces(1)),
+                                   expression: ExprSyntax(StringLiteralExprSyntax(package.displayName)),
+                                               trailingComma: nil)
+        ])
+
+        let callExpr = FunctionCallExprSyntax(calledExpression: ExprSyntax(dotProductExpr),
+                                                          leftParen: TokenSyntax.leftParenToken(),
+                                                          argumentList: argumentList,
+                                                          rightParen: TokenSyntax.rightParenToken(),
+                                                          trailingClosure: nil,
+                                                          additionalTrailingClosures: nil)
+
+        // Add the target dependency entry.
+        let newManifest = targetDependencies.withAdditionalElementExpr(ExprSyntax(callExpr)).root
+        self.editedSource = newManifest.as(SourceFileSyntax.self)!
+    }
+
+    private func findTargetDependenciesArrayExpr(target: String) throws -> ArrayExprSyntax {
+        let initFnExpr = try findPackageInit()
+
+        // Find the `targets: []` array.
+        let targetsArrayFinder = ArrayExprArgumentFinder(expectedLabel: "targets")
+        targetsArrayFinder.walk(initFnExpr.argumentList)
+        guard case .found(let targetsArrayExpr) = targetsArrayFinder.result else {
+            diagnosticsEngine.emit(.missingPackageInitArgument(name: "targets"))
+            throw Diagnostics.fatalError
```

Comment:
should this throw instead of editing diagnostics then throwing generic diagnostics error?